paddle/fluid/framework/new_executor/data_transfer.cc

```diff
@@ -110,33 +112,61 @@ void DataTranferHelper::RunAndConstructOpFuncNode(
   runtime_context.inputs["X"] = {scope_->FindVar(var_name)};
   runtime_context.outputs["Out"] = {scope_->Var(new_var_name)};
   InterpretercoreInferShapeContext infer_shape_ctx(*op, runtime_context);
-
-  // 2. Execute infer shape and choose kernel
-  auto& all_op_kernels = OperatorWithKernel::AllOpKernels();
   op.get()->Info().infer_shape_(&infer_shape_ctx);
-  auto kernels_iter = all_op_kernels.find(op_type);
-  PADDLE_ENFORCE_NE(kernels_iter,
-                    all_op_kernels.end(),
-                    platform::errors::Unavailable(
-                        "There are no kernels which are registered in "
-                        "the %s operator.",
-                        op_type));
-  OpKernelMap& kernels = kernels_iter->second;
+
+  // 2. choose kernel
+
+  // prepare a ptr to OperatorWithKernel
+  OperatorBase* op_ptr = op.get();
+  if (dynamic_cast<framework::OperatorWithKernel*>(op_ptr) == nullptr) {
+    PADDLE_THROW(platform::errors::PreconditionNotMet(
+        "%s should be OperatorWithKernel type.", op_ptr->Type()));
+  }
+  auto op_with_kernel = static_cast<framework::OperatorWithKernel*>(op_ptr);
+
   platform::DeviceContextPool& pool = platform::DeviceContextPool::Instance();
   auto* dev_ctx = pool.Get(place_);
-  Scope scope;
-  auto exec_ctx = ExecutionContext(*op, scope, *dev_ctx, runtime_context);
-  auto expected_kernel_key =
-      dynamic_cast<const framework::OperatorWithKernel*>(op.get())
-          ->GetExpectedKernelType(exec_ctx);
-  auto kernel_iter = kernels.find(expected_kernel_key);
+  auto exec_ctx = ExecutionContext(*op, Scope(), *dev_ctx, runtime_context);
+  auto expected_kernel_key = op_with_kernel->GetExpectedKernelType(exec_ctx);
+
+  VLOG(6) << "expected_kernel_key " << expected_kernel_key << "\n";
+  VLOG(6) << "op_with_kernel Type() " << op_with_kernel->Type() << "\n";
+
+  bool run_phi_kernel = false;
+
+  // check if phi kernel exists
+  auto phi_kernel_map =
+      phi::KernelFactory::Instance().SelectKernelMap(op_with_kernel->Type());
+  if (phi_kernel_map.size() > 0) {
+    auto phi_kernel_key = op_with_kernel->ChoosePhiKernel(exec_ctx);
+    VLOG(6) << "phi_kernel_key " << phi_kernel_key << "\n";
+
+    PADDLE_ENFORCE_EQ(op_with_kernel->PhiKernel()->IsValid(),
+                      true,
+                      platform::errors::PreconditionNotMet(
+                          "this function was expected to construct"
+                          "data transfer op, but %s was passed.",
+                          op_with_kernel->Type()));
```

Comment:
To make it clear. -> "The xx op has no valid phi kernel"

Reply:
done